jq/private/rules/eval.bzl

```diff
@@ -0,0 +1,119 @@
+"jq_eval rule"
```

Comment:
What's `jq_eval` for?  Does anything use it outside this repo?

Reply:
I'll need to add documentation to expand this (there's a bazel-convention for documentation) but it kinda goes like this:

1. Bazel files are kinda like `Makefiles` but without so much of a special DSL
   - Bazel config (.bzl) files are written in "starlark", which is like a subset of python
   - import statements are `load(...)` statements
1. a "toolchain" is a mapping to a binary for a specific platform (os, architecture)
   - likely gets it's name from making a cc compiler available
   - recall that jq-1.6 didn't have ARM binaries, so it won't be cool for ARM yet (it will with 1.7)
1. Buck (Facebook) is similar (but different imports and more apache-thrift)
   - Bazel and Buck are based on Blaze, a google-internal project
   
 Essentially `jq_eval()` here is a library function; if you look at `example/eval/BUILD.bazel` you'll see:
 ```
 load("//jq:defs.bzl", "jq_eval")

 jq_eval(
     name = "actual",
     srcs = ["sample.yaml"],
     filter = ".a.b",
     raw_output = True,
 )
```
The `load(` is like a `from devs.bzl import jq_eval` in python.  The `jq_eval()` later exercises that command, going back to the `def jq_eval()` macro to do the work, if it's defined as a python-style macro.  If it's a `jq_eval = rule (...` then there's a different set of possibility: different default values, more rigor, ability to use a toolchain, or the make-style variables (which is checked in `example/genrule/`).

This is effectively a `jq .a.b example/eval/sample.yaml`, confirmed with a unittest:
```
diff_test(
     name = "check_eval"
    ...
)
```

If you type `bazel build //example/eval:actual`, bazel will track down a `jq` binary corresponding to your platform (OS, architecture), download it and make it available.

If you type `bazel build //...`, "//..." being a wildcard-ish statement, bazel will build everything not flagged as "manual" ie "don't build by wildcard but vy dependency or specific mention".

If you type `bazel test //...` or `bazel test //example/eval:check_eval`, bazel will exercise that test that confirms that `jq .a.b sample.yaml` produces `123`.

I can transpose this to documentation in a following PR.

Reply:
> Does anything use it outside this repo?


I think most use of `jq` in other Bazel projects is by the form similar to `example/genrule`: the `JQ_BIN` population by the underlying toolchain.  Other rules can then use `$(JQ_BIN)` like a makefile variable, and the JQ binary is only pulled on-demand, ie when the rule that needs it is run (pulled from local cache, local repository cache, federated cache, and caching-proxy where available).